### PR TITLE
Add byte-size limit to transaction & block txids' caches

### PR DESCRIPTION
**Pull request by romanz**
Also, move TransactionCache into `cache` module.

Following #161 by @dagurval.

### CODE REVIEW

File: src/cache.rs [hunk 1]
```diff
@@ -34,29 +80,93 @@ impl BlockTxIDsCache {
...
+    pub fn get_or_else<F>(&self, txid: &Sha256dHash, load_txn_func: F) -> Result<Transaction>
+    where
+        F: FnOnce() -> Result<Transaction>,
+    {
+        if let Some(txn) = self.map.lock().unwrap().get(txid) {
+            return Ok(txn.clone());
+        }
+        let txn = load_txn_func()?;
+        let byte_size = 32 /* key (hash size) */ + serialize(&txn).len();
+        self.map.lock().unwrap().put(*txid, txn.clone(), byte_size);
```
[the remark this answers is not in the document]
Comment (romanz) [reply src/cache.rs]: Good point, thanks!
Fixed at 590353a9545f5ff5150c53010fd5f39c8d95aa4a.

File: src/cache.rs [hunk 2]
```diff
@@ -1,27 +1,73 @@
...
+                None
+            }
+            Some((value, _)) => {
+                self.lookups.with_label_values(&["hit"]).inc();
+                Some(value)
+            }
+        }
+    }
+
+    fn put(&mut self, key: K, value: V, byte_size: usize) {
```
[the remark this answers is not in the document]
Comment (romanz) [reply src/cache.rs]: Good point, thanks!
Fixed at 590353a9545f5ff5150c53010fd5f39c8d95aa4a.

**Comment by dagurval** [comment]
Except for the `unwrap` nit, the changes look good :-)

File: src/cache.rs [hunk 3]
```diff
@@ -34,29 +83,110 @@ impl BlockTxIDsCache {
...
+        F: FnOnce() -> Result<Vec<u8>>,
+    {
+        match self.map.lock().unwrap().get(txid) {
+            Some(serialized_txn) => Ok(deserialize(&serialized_txn).unwrap()),
+            None => {
+                let serialized_txn = load_txn_func()?;
+                let txn = deserialize(&serialized_txn).chain_err(|| "failed to parse cached tx")?;
+                let byte_size = 32 /* key (hash size) */ + serialized_txn.len();
+                self.map
+                    .lock()
```
Comment (dagurval) [inline src/cache.rs]: After testing, I found that there is recursive locking happening here and the software hangs.

This patch fixed it for me:
```
--- a/src/cache.rs
+++ b/src/cache.rs
@@ -120,18 +120,17 @@ impl TransactionCache {
         F: FnOnce() -> Result<Vec<u8>>,
     {
         match self.map.lock().unwrap().get(txid) {
-            Some(serialized_txn) => Ok(deserialize(&serialized_txn).unwrap()),
-            None => {
-                let serialized_txn = load_txn_func()?;
-                let txn = deserialize(&serialized_txn).chain_err(|| "failed to parse cached tx")?;
-                let byte_size = 32 /* key (hash size) */ + serialized_txn.len();
-                self.map
-                    .lock()
-                    .unwrap()
-                    .put(*txid, serialized_txn, byte_size);
-                Ok(txn)
-            }
+            Some(serialized_txn) => return Ok(deserialize(&serialized_txn).unwrap()),
+            None => {}
         }
+        let serialized_txn = load_txn_func()?;
+        let txn = deserialize(&serialized_txn).chain_err(|| "failed to parse cached tx")?;
+        let byte_size = 32 /* key (hash size) */ + serialized_txn.len();
+        self.map
+            .lock()
+            .unwrap()
+            .put(*txid, serialized_txn, byte_size);
+        Ok(txn)
     }
```

Comment (romanz) [reply src/cache.rs]: Many thanks for finding and fixing the issue!
Force-pushed https://github.com/romanz/electrs/pull/172/commits/51d0bf4fb77eb7d8354e3ce2f6de7a9b0c6dcb54.

**Comment by romanz** [comment]
Thanks for the contribution!

**Pull request merged**